### PR TITLE
[knx] Mods to enable the KNX ‘ETS’ Binding

### PR DESCRIPTION
**Pull request by kgoderis**
This PR contains all the changes to the KNX Basic binding that are required by https://github.com/openhab/openhab2-addons/pull/3451. 

Signed-Off-By: Karel Goderis <karel.goderis@me.com>

### CODE REVIEW

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+    public final Set<String> getChannelIDs() {
+        return channelTypeIDs;
+    }
+
+    protected final Set<String> asSet(String... values) {
+        return Collections.unmodifiableSet(new HashSet<>(Arrays.asList(values)));
+    }
+
+    protected abstract @NonNull Set<@NonNull String> getAllGAKeys();
```
Comment (wborn) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java]: You should annotate the class with `@NonNullByDefault` and use `@Nullable` where needed. 

See also the [ESH documentation](https://www.eclipse.org/smarthome/documentation/development/conventions.html#null-annotations).

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java [hunk 2]
```diff
@@ -103,6 +113,11 @@
     /** stores the default KNX DPT to use for each openHAB type */
     private final Map<Class<? extends Type>, String> defaultDptMap;
 
+    /** stores the default Item Type to use for each openHAB type */
+    private final Map<Class<? extends Type>, Class<? extends GenericItem>> defaulItemTypeMap;
+
+    private HashMap<Class<? extends GenericItem>, String> itemTypeStringMap;
```
Comment (wborn) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java]: Can you change the type from HashMap to Map here?

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java [hunk 3]
```diff
@@ -116,6 +131,9 @@ public KNXCoreTypeMapper() {
 
         dptTypeMap = new HashMap<String, Class<? extends Type>>();
         dptMainTypeMap = new HashMap<Integer, Class<? extends Type>>();
+        defaultDptMap = new HashMap<Class<? extends Type>, String>();
+        defaulItemTypeMap = new HashMap<Class<? extends Type>, Class<? extends GenericItem>>();
+        itemTypeStringMap = new HashMap<Class<? extends GenericItem>, String>();
```
Comment (wborn) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java]: Can you use the [diamond operator](https://www.javaworld.com/article/2074080/core-java/jdk-7--the-diamond-operator.html) here?

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/LoadState.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.knx.internal.handler;
+
+/**
+ * Enumeration containing the load states
+ *
+ * @author Karel Goderis - Initial contribution
+ */
+public enum LoadState {
```
Comment (wborn) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/LoadState.java]: Is this Enum used? I couldn't find any references to it.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/LoadState.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    L0(0, "Unloaded"),
+    L1(1, "Loaded"),
+    L2(2, "Loading"),
+    L3(3, "Error"),
+    L4(4, "Unloading"),
+    L5(5, "Load Completing");
+
+    private int code;
+    private String name;
```
Comment (wborn) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/LoadState.java]: @kgoderis I was searching for uses of this Enum to find out how the `name` property was used. The `name` property and `getName(..)` methods of this Enum are confusing because an Enum already has a `.name()` method. Maybe the property can be renamed to something else so there will be no confusion?

**Comment by kgoderis** [comment]
@sjka Simon, any thoughts on this PR?

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java [hunk 6]
```diff
@@ -832,6 +868,19 @@ public String toDPTid(Class<? extends Type> typeClass) {
...
+    @Override
+    public Class<? extends GenericItem> toItemTypeClass(Class<? extends Type> typeClass) {
+        return defaulItemTypeMap.get(typeClass);
+    }
+
+    public Class<? extends GenericItem> toItemTypeClass(String dpt) {
+        return defaulItemTypeMap.get(toTypeClass(dpt));
+    }
+
+    public String toItemType(String dpt) {
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java]: This is neither part of the public API (i.e. interface) nor is it used internally. Therefore this method is dead code and can be deleted.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java [hunk 7]
```diff
@@ -832,6 +868,19 @@ public String toDPTid(Class<? extends Type> typeClass) {
         return defaultDptMap.get(typeClass);
     }
 
+    @Override
+    public Class<? extends GenericItem> toItemTypeClass(Class<? extends Type> typeClass) {
+        return defaulItemTypeMap.get(typeClass);
+    }
+
+    public Class<? extends GenericItem> toItemTypeClass(String dpt) {
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java]: dead code - see below

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java [hunk 8]
```diff
@@ -534,6 +551,25 @@ public KNXCoreTypeMapper() {
         defaultDptMap.put(DateTimeType.class, DPTXlatorTime.DPT_TIMEOFDAY.getID());
         defaultDptMap.put(StringType.class, DPTXlatorString.DPT_STRING_8859_1.getID());
         defaultDptMap.put(HSBType.class, DPTXlatorRGB.DPT_RGB.getID());
+
+        defaulItemTypeMap.put(DateTimeType.class, DateTimeItem.class);
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java]: You are effectively duplicating (partially) the information from acceptedItemTypes/acceptedCommandTypes here. Not sure if that's a good idea...

File: addons/binding/org.openhab.binding.knx/META-INF/MANIFEST.MF [hunk 9]
```diff
@@ -11,6 +11,8 @@ Import-Package: gnu.io;resolution:=optional,
  org.eclipse.smarthome.config.discovery,
  org.eclipse.smarthome.core.common,
  org.eclipse.smarthome.core.i18n,
+ org.eclipse.smarthome.core.items,
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/META-INF/MANIFEST.MF]: I think we had this discussion a couple of times already: Bindings must be agnostic of items. Otherwise the whole functional abstraction won't work. So I hope you'll understand this won't pass without a single word of explanation?

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    protected final Set<String> channelTypeIDs;
+
+    private final Logger logger = LoggerFactory.getLogger(AbstractKNXChannelType.class);
+
+    protected AbstractKNXChannelType(String... channelTypeIDs) {
+        this.channelTypeIDs = new HashSet<>(Arrays.asList(channelTypeIDs));
+    }
+
+    @Nullable
+    protected abstract ChannelConfiguration parse(@Nullable String fancy);
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java]: The `ChannelConfiguration` class is something internal to the "basic" KNX binding and therefore must not "leak" out of the API.

Most of the methods below have the same "issue": They are specific to the way the basic binding configures channels and therefore cannot reside in a commonly shared abstract class.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.knx.internal.channel;
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java]: If you plan to extend it, then this must not be in an "internal" package. I don't think though this is a good idea at all (see below).

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractSpec.java [hunk 12]
```diff
@@ -43,9 +43,13 @@ protected AbstractSpec(@Nullable ChannelConfiguration channelConfiguration, Stri
...
      * @return a group address object
      */
-    protected final GroupAddress toGroupAddress(GroupAddressConfiguration ga) {
+    protected final GroupAddress toGroupAddress(@Nullable GroupAddressConfiguration ga) {
         try {
-            return new GroupAddress(ga.getGA());
+            if (ga != null) {
+                return new GroupAddress(ga.getGA());
+            } else {
+                throw new IllegalArgumentException("GroupAddressConfiguration should contain a group address");
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractSpec.java]: Why did you change the signature to be `@Nullable` and then throw a IAE in case `null` is passed?

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/ChannelConfiguration.java [hunk 13]
```diff
@@ -26,10 +27,10 @@
 public class ChannelConfiguration {
 
     private final @Nullable String dpt;
-    private final GroupAddressConfiguration mainGA;
+    private final @Nullable GroupAddressConfiguration mainGA;
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/ChannelConfiguration.java]: No, that's a pretty fundamental assumption: the first one is the "main" GA - and IF there is a configuration, it MUST contain an address.

Are you maybe changing the meaning of "main GA"? Then please don't do so for the basic binding.

Actually, I don't think it is a good idea to reuse this class. Advanced binding(s) may make their own choices how a channels is configured. That's why this is in an "internal" package - it is _internal_ to the basic binding.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+                String dpt = config.getDPT();
+                if (dpt == null) {
+                    dpt = getDefaultDPT(key);
+                }
+
+                Class<? extends Type> expectedTypeClass = typeHelper.toTypeClass(dpt);
+                if (expectedTypeClass != null) {
+                    logger.trace("The expected Type class for dpt '{}' is '{}'", dpt,
+                            expectedTypeClass.getSimpleName());
+                    Type convertedType = convertType(command, configuration);
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java]: How is this related to the "ETS" binding? Aren't you changing the basic binding's behavior here?

Don't get me wrong, potentially this makes sense. I'm just not quite sure yet about the side-effects this might have. In any case, I'm just trying to understand the reason for this change.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/WriteSpecImpl.java [hunk 15]
```diff
@@ -26,12 +28,13 @@
     private final Type type;
     private final @Nullable GroupAddress groupAddress;
 
+    private final Logger logger = LoggerFactory.getLogger(WriteSpecImpl.class);
```
Comment (sjka) [inline addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/WriteSpecImpl.java]: what do you need this one for?

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java (hunk 6 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java]: It is used in the advanced binding, and in that context its natural place is in the KNXCoreTypeMapper (it can be potentially be used by other knx binding derivaties), but if it is really an issue I will put it in an inheriting class.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java (hunk 8 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java]: In the advanced binding I do implement I use a ChannelBuilder to create Channels based on the information stored in the ETS5 files. ChannelBuilder::create() takes an acceptedItemType as input, which is provided by the defaulItemTypeMap you mention, and for which in turn the Key is provided by KNXCoreTypeMapper::toTypeClass() based on the DPT that is derived from the ETS5 data.

File: addons/binding/org.openhab.binding.knx/META-INF/MANIFEST.MF (hunk 9 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.knx/META-INF/MANIFEST.MF]: See https://github.com/openhab/openhab2-addons/pull/3500#discussion_r189028969

I do not use Items directly, but I need their class definitions for the TypeMapper.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java (hunk 10 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java]: My main objective was to re-use as much as possible existing classes without re-inventing the wheel. The only part that is specific for each binding is the parsing of the configuration String. In fact, in the advanced binding I do construct a configuration String based on ETS5 information   that is similar (but not the same) as the basic binding's configuration Strings

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractSpec.java (hunk 12 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractSpec.java]: I should revert this and do the null check before calling this. Underlying to this is that in the advanced binding there can be ChannelConfigurations that do not have a MainGA (toGroupAddress is called in ReadResponseSpecImpl)

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/ChannelConfiguration.java (hunk 13 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/ChannelConfiguration.java]: mmh.. It is related to https://github.com/openhab/openhab2-addons/pull/3500#discussion_r189029968 . Maybe some of the internal classes of the basic binding should be exposed in order to facilitate re-usage elsewhere.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java (hunk 14 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java]: For the basic binding this function returns the Type is was fed without any modification. In the advanced binding I overload it to do something meaningful. See for example TypeRollerShutter.java on how I do use this functionality.

File: addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/WriteSpecImpl.java (hunk 15 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/WriteSpecImpl.java]: Debugging artefact. It should be removed indeed

**Comment by lewie** [comment]
@kgoderis, the review of sjka is complete. I personally think that without extensions in the basic binding, we will not get along. Duplication of source code is much more complicated and error-prone.

I would suggest to mark the parts required for extensions. A short Comment with the reason in the places as a dead code appear and in the classes which should not be internal. 

That would be the easiest way to get ahead?
@sjka, would that be a solution for your comments?

**Comment by vguna** [comment]
Any news on this? This would be a really nice feature for the KNX integration into OH!

**Comment by Hilbrand** [comment]
Given the discussion here it seems a valuable addition. But due too long time no activity closing this pr. if someone wants to pick this up please open a new pr.
